lua/script.lua

```diff
@@ -3,106 +3,69 @@
 
 local Script = {}
 
---- reset script environment;
--- ie redirect draw, key, enc functions, stop timers, clear engine, etc
-Script.clear = function()
-  print("# script clear")
-  -- reset cleanup script
-  cleanup = norns.none
-  -- reset oled redraw
-  redraw = norns.blank
-  -- redirect inputs to nowhere
-  key = norns.none
-  enc = norns.none
-  -- clear, redirect, and reset grid
-  if g then 
-    g:all(0)
-    g:refresh()
-    g.key = norns.none
-  end
-  g = nil
-  -- reset gridkey callback
-  gridkey = norns.none
-  -- reset midi callbacks
-  midi.add = norns.none
-  midi.remove = norns.none
-  -- stop all timers
-  metro.free_all()
-  -- stop all polls and clear callbacks
-  poll.clear_all()
-  -- clear engine
-  engine.name = nil
-  -- clear init
-  init = norns.none
-  -- clear last run
-  norns.state.script = ''
-  norns.state.name = 'none'
-  -- clear params
-  params:clear()
-end
+Script.file = ''
+Script.name = 'none' 
+Script.errorstate = false
+Script.errormsg = ''
 
-Script.init = function()
+--- run script init() function with error checking
+function Script.init()
   print("# script init")
-  params.name = norns.state.name
-  if norns.try(init, "init") then
-    norns.menu.init()
-  end
+  params.name = norns.script.name
+  norns.try(init, "init")
+  norns.init_done()
 end
 
---- load a script from the /scripts folder
+--- load a script from the /scripts folder, reset LVM
 -- @param filename (string) - file to load. leave blank to reload current file.
-Script.load = function(filename)
+function Script.load(filename)
   print("# script load")
   if filename == nil then
-    filename = norns.state.script end
+    filename = Script.file end
   local filepath = script_dir .. filename
   local f=io.open(filepath,"r")
   if f==nil then
     print("file not found: "..filepath)
+    Script.file = ''
   else
     io.close(f)
-    if pcall(cleanup) then print("# cleanup") 
-    else print("### cleanup failed") end
-
-    Script.clear() -- clear script variables and functions
-    
-    local status = norns.try(function() dofile(filepath) end, "load fail") -- do the new script
-    if status == true then
-      norns.log.post("loaded " .. filename) -- post to log
-      norns.state.script = filename -- store script name
-      norns.state.folder_name = string.gsub(filename,'.lua','') -- store name
-      norns.state.name = norns.state.folder_name:match("[^/]*$") -- strip path from name
-      norns.state.save() -- remember this script for next launch
-      norns.script.redraw = redraw -- store redraw function for context switching
-      redraw = norns.none -- block redraw until Script.init
-      Script.run() -- load engine then run script-specified init function
-    end
+    Script.file = filename
   end
+  if pcall(cleanup) then print("# cleanup") 
```

Comment:
maybe `norns.try` or something similar here?  notice that `norns.try` uses `xpcall` which gives us better error tracing:

https://github.com/monome/norns/blob/4e908e3edeb1ce5f49efc6d072aadbc0dd6fe1d6/lua/norns.lua#L160